accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -864,9 +868,19 @@ _is the extension type_
 <code>V\<T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>,
 and that its static type _is an extension type_.
 
-It is a compile-time error if `await e` occurs, and the static type of
-`e` is an extension type which is not a subtype of `Future<T>` for any
-`T`.
+We say that a type `T` is _incompatible with await_ if at least 
+one of the following criteria holds:
+
+- `T` is an extension type, and `T` does not implement `Future`.
+- `T` is `S?`, and `S` is incompatible with await.
+- `T` is `X & B`, and either:
+  - `B` is incompatible with await, or
+  - `B` does not derive a future type, and `X` is
+    incompatible with await.
+- `T` is `S` bounded, and `S` is incompatible with await.
```

Comment:
I still prefer to keep every rule one-step, so this one would be
> - `T` is a type variable `X` with bound `S`, and `S` is incompatible with await.

By using "`S` bounded" we include both transitive bounds *and* promoted type variables, which means that "`X & B` where `B` is incompatible with await" is now double-handled, since it's *also* handled by "`S` bounded and `S` is incompatible with await".

It's probably not wrong, but it introduces an extra complexity that makes it harder to reason about what the cases actually do. It's not just the "T is a type variable with bound `S`" or even just the transitive of that, because it also includes promoted type variables.
And when it's the only rule which does that, it stands out and makes the rest of the rules harder to think about too, when you can't use the same approach to every rule.

If we ever change "`S`-bounded" to include something more, we risk this rule being affected, even if we don't 
actually *want* it. It probably won't, but I'd rather be explict.

Reply:
OK, changed the rule to use your wording.

I was a little bit worried about reducing the coverage of the ruleset as a whole (such that we don't catch situations that we would catch with "bounded"), but I don't see any such cases with the concrete set of examples that we've discussed, and a check on the definition of "bounded" seems to support the claim that nothing is lost.

This is great, the rules are actually rather easy to read now! ;-)